content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,353 @@
+`std::atomic` and `std::atomic_thread_fence`.
+
+```C++
+class Foo {
+    private:
+        std::atomic <Foo*> helper;
+    public:
+        Foo* get_helper() {
+            Foo* h = helper.load(std::memory_order_relaxed);
+            std::atomic_thread_fence(std::memory_order_acquire);        //memory barrier
+            if (h == nullptr) {
+                std::lock_guard<std::mutex> lock(m_init);
+                h = helper.load(std::memory_order_relaxed);
+                if (h == nullptr) {
+                    h = new Helper;
+                    std::atomic_thread_fence(std::memory_order_release);//memory barrier
+                    helper.store(h, std::memory_order_relaxed);
+                }
+            }
+            return tmp;
+        }
+};
+```
+
+* `std::atomic_thread_fence(std::memory_order_acquire)` guarantees that 
+read/write after a memory barrier cannot be reordered before the memory barrier.
+* `std::atomic_thread_fence(std::memory_order_release)` guarantees that 
+read/write before a memory barrier cannot be reordered after the memory barrier.
+
+Similar to the Java volatile case, 
+the initialization must be done before it stores `h` to `helper` 
+with the memory barrier.
+
+### Static Singleton
+
+If the `helper` is static, i.e. all the instances of class `Foo` shares the 
+same instance of `helper`, defining the `helper` in a static field of a separate
+class will solve the problem.
+
+```Java
+class Foo {
+    private static class HelperSingleton {
+        public static final Helper helper = new Helper();
+    }
+
+    public Helper getHelper() {
+        return HelperSingleton.helper;
+    }
+}
+```
+
+This is known as **initialization-on-demand holder idiom**, 
+which is a safe and efficient concurrent lazy initialization 
+for all Java versions.
+
+* **Lazy Initialization**: as specified by the Java language,
+    the static class `HelperSingleton` is not initialized until the first it is
+    referenced.
+* **Safe**: the class initialization is guaranteed to be synchronized by 
+    Java language.
+* **Efficient**: all subsequent calls of `getHelper` will return the initialized
+    `helper` without any synchronization overhead.
+
+### Thread Local
+
+`ThreadLocal` is a variable that each thread will have its own copy of its value.
+Each thread can only access and modify its own copy of a thread local variable
+independently of other threads.
+
+A thread local can be used to maintain the state of "whether the state has gone
+through the synchronized initialization". 
+Since if a thread has gone through the synchronized initialized once,
+it can be confident about that object is already initialized.
+
+Inside the synchronized initialization section, 
+only the first thread will find the object is `null` and initialized object.
+All threads will change their per-thread state, 
+    so that they will not enter the synchronized section again.
```

Comment:
Maybe it's worth expanding on the fact that _every_ thread will then execute the critical section _exactly once_? That's clearly more costly than the "ideal" double-checked locking pattern, if it were to work. But it's not much worse, so that's probably fine.